packages/schedule/src/Schedule.js

```diff
@@ -3,174 +3,327 @@
+  firstCallbackNode = node;
+
+  // Remove the node from the list before calling the callback. That way the
+  // list is in a consistent state even if the callback throws.
+  const next = firstCallbackNode.next;
+  if (firstCallbackNode === next) {
+    // This is the last callback in the list.
+    firstCallbackNode = null;
+  } else {
+    const previous = firstCallbackNode.previous;
+    firstCallbackNode = previous.next = next;
+    next.previous = previous;
+  }
+
+  node.next = node.previous = null;
+
+  // Now it's safe to call the callback.
+  const callback = node.callback;
+  callback(deadlineObject);
+}
+
+function flushWork(didTimeout) {
+  isPerformingWork = true;
+  deadlineObject.didTimeout = didTimeout;
+  try {
+    if (firstCallbackNode !== null) {
+      if (didTimeout) {
+        // Flush all the timed out callbacks without yielding.
+        do {
+          flushCallback(firstCallbackNode);
+        } while (
+          firstCallbackNode !== null &&
+          firstCallbackNode.timesOutAt <= getCurrentTime()
+        );
+      } else {
+        // Keep flushing callbacks until we run out of time in the frame.
+        while (
+          firstCallbackNode !== null &&
+          getTimeRemaining() - getCurrentTime() > 0
+        ) {
+          flushCallback(firstCallbackNode);
+        }
+      }
+    }
+  } finally {
+    isPerformingWork = false;
+    if (firstCallbackNode !== null) {
+      // There's still work remaining. Request another callback.
+      ensureHostCallbackIsScheduled(firstCallbackNode);
+    } else {
+      isHostCallbackScheduled = false;
+    }
+  }
+}
+
+function unstable_scheduleWork(callback, options) {
+  const currentTime = getCurrentTime();
+
+  let timesOutAt;
+  if (options !== undefined && options !== null) {
+    const timeoutOption = options.timeout;
+    if (timeoutOption !== null && timeoutOption !== undefined) {
+      // If an explicit timeout is provided, it takes precedence over the
+      // priority context.
+      timesOutAt = currentTime + timeoutOption;
+    } else {
+      // Compute an absolute timeout using the current priority context.
+      timesOutAt = computeAbsoluteTimeoutForPriority(
+        currentTime,
+        priorityContext,
+      );
+    }
+  } else {
+    timesOutAt = computeAbsoluteTimeoutForPriority(
+      currentTime,
+      priorityContext,
+    );
+  }
+
+  const newNode = {
+    callback,
+    timesOutAt,
+    next: null,
+    previous: null,
+  };
+
+  // Insert the new callback into the list, sorted by its timeout.
+  if (firstCallbackNode === null) {
+    // This is the first callback in the list.
+    firstCallbackNode = newNode.next = newNode.previous = newNode;
+    ensureHostCallbackIsScheduled(firstCallbackNode);
+  } else {
+    let next = null;
+    let node = firstCallbackNode;
+    do {
+      if (node.timesOutAt > timesOutAt) {
+        // This callback is lower priority than the new one.
```

Comment:
I'd like to avoid using the word "priority" in this file if we can since mixing that with timeouts changes the meaning of "higher" and "lower".

e.g.

```js
// The new callback times out before this one
```